### ISSUE TITLE
CSI Backup fails upon provisioner and csi driver mismatch (everest-csi)

### ISSUE THREAD

**Issue by lestich** [vmware-tanzu/velero#8277]
**What steps did you take and what happened:**
I tried to create a velero CSI backup on my cluster, running on OpenTelekomCloud using everest-csi as storage provider, but it didn't work. VolumeSnapshots are supported and work as expected when i create them via VolumeSnapshot CR. Velero cannot handle the mismatch between StorageClass provisioner and driver in VolumeSnapshotClass ( see down below ). I tried to set the VolumeSnapshotClass explicit via PVC annotation but it fails upon this check:

https://github.com/vmware-tanzu/velero/blob/release-1.14/pkg/util/csi/volume_snapshot.go#L353

Logs:

"Didn't find VolumeSnapshotClass from PVC annotations: Incorrect VolumeSnapshotClass csi-disk-snapclass is not for driver everest-csi-provisioner" backup=velero/test-annotation3 cmd=/velero logSource="pkg/util/csi/volume_snapshot.go:313" pluginName=velero

"Error backing up item" backup=velero/test-annotation3 error="error executing custom action (groupResource=persistentvolumeclaims, namespace=default, name=hello-world-pvc): rpc error: code = Unknown desc = failed to get VolumeSnapshotClass for StorageClass csi-disk-topology-default: error getting VolumeSnapshotClass: failed to get VolumeSnapshotClass for provisioner everest-csi-provisioner, \n\t\tensure that the desired VolumeSnapshot class has the velero.io/csi-volumesnapshot-class label

**What did you expect to happen:**
Setting the VSC Annotation via the PVC should lead to a functional backup (by ignoring the mismatch between provisioner and driver)

**Anything else you would like to add:**
storageclass used:
`
allowVolumeExpansion: true
apiVersion: storage.k8s.io/v1
kind: StorageClass
metadata:
  annotations:
    storageclass.kubernetes.io/is-default-class: "true"
  name: csi-disk-topology-default
parameters:
  csi.storage.k8s.io/csi-driver-name: disk.csi.everest.io
  csi.storage.k8s.io/fstype: ext4
  everest.io/disk-volume-type: SAS
  everest.io/passthrough: "true"
provisioner: everest-csi-provisioner
reclaimPolicy: Delete
volumeBindingMode: WaitForFirstConsumer
`

volumesnapshotclass used:
`
apiVersion: snapshot.storage.k8s.io/v1
deletionPolicy: Delete
driver: disk.csi.everest.io
kind: VolumeSnapshotClass
metadata:
  annotations:
    snapshot.storage.kubernetes.io/is-default-class: "true"
  labels:
    velero.io/csi-volumesnapshot-class: "true"
  name: otc-vsc
`
VolumeSnapshots work as expected with this (strange) configuration. It is preprovisioned by the cloud-provider / everest-csi addon (beside the annotations for default-class and velero label).

**Environment:**

- Velero version (use `velero version`): v.1.14.1
- Velero features (use `velero client config get features`): EnableCSI
- Kubernetes version (use `kubectl version`): 1.29

**Vote on this issue!**

This is an invitation to the Velero community to vote on issues, you can see the project's [top voted issues listed here](https://github.com/vmware-tanzu/velero/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc).  
Use the "reaction smiley face" up to the right of this comment to vote.

- :+1: for "I would like to see this bug fixed as soon as possible"
- :-1: for "There are more important bugs to focus on right now"

**Comment by msfrucht** [comment]
Applications that automatically choose a VolumeSnapshotClass for snapshots are supposed to use StorageClass provisioner field to match or at least contain the same value in VolumeSnapshotClass driver field. And this is shown in the everest snapshot examples as well.

https://github.com/huaweicloud/huaweicloud-csi-driver/tree/master/examples/evs-csi-plugin/kubernetes/snapshot

Change these fields to match and the issue will be resolved.

**Comment by ywk253100** [comment]
@lestich Is it possible to modifier the `StorageClass` and `VolumeSnapshotClass` to make them match as @msfrucht suggested?

**Comment by lestich** [comment]
I tried that yesterday already . If i change the provisioner in StorageClass to disk.csi.everest.io it cannot provision PVs anymore and if i change the driver in VolumeSnapshotClass to everest-csi Snapshots are not created (in this case velero doesn't fail early but times out)

**Comment by msfrucht** [comment]
That is odd. A CSI driver has to register a provisioner string. That implies that when it was installed the provisioner string was set to everest-csi-provisioner instead of the default "evs.csi.huaweicloud.com".

It should be far safer to adjust the VolumeSnapshotClass than the StorageClass.

**Comment by lestich** [comment]
Any updates on this ? Do you need more information ? 

As a workaround we forked the velero project and removed this check:
https://github.com/vmware-tanzu/velero/blob/release-1.14/pkg/util/csi/volume_snapshot.go#L353

After that it works. Are there any plans to modify this check i.e. to use a fixed VolumeSnapshotClass via annotation ?

**Comment by lestich** [comment]
> Applications that automatically choose a VolumeSnapshotClass for snapshots are supposed to use StorageClass provisioner field to match or at least contain the same value in VolumeSnapshotClass driver field. And this is shown in the everest snapshot examples as well.
> 
> https://github.com/huaweicloud/huaweicloud-csi-driver/tree/master/examples/evs-csi-plugin/kubernetes/snapshot
> 
> Change these fields to match and the issue will be resolved.

You are right, the huaweicloud-csi-driver does this correct. Unfortunately we are using the OTC cloud which seems to use a proprietary version of this (everest-csi-provisioner instead of evs.csi.huaweicloud.com)

**Comment by SebastianOpeni** [comment]
We sadly also encounter this problem.
So for now we either have to also fork the project or try to utilize a different backup-solution. :C

**Comment by weeix** [comment]
I believe that explicitly annotating the PVC with the desired VolumeSnapshotClass should bypass the provisioner and driver mismatch check. This would prevent errors and enhance the user experience.

If the check must remain, a clearer error message would help. The current message, "failed to get VolumeSnapshotClass for provisioner everest-csi-provisioner, ensure that the desired VolumeSnapshot class has the velero.io/csi-volumesnapshot-class label," is misleading, as it suggests the label is missing when the VolumeSnapshotClass is explicitly specified.

Also, the documentation at https://velero.io/docs/main/csi/#implementation-choices does not clarify the need for the StorageClass provisioner to match the VolumeSnapshotClass driver, adding to the confusion.

**Comment by kaovilai** [comment]
Feel free to PR the clearer error message.

**Comment by livelace** [comment]
Any news to fix this check/behavior ?

**Comment by lubronzhan** [comment]
https://github.com/velero-io/velero/pull/9214
This was fixed and can be closed and it's part of 1.18
